File: kyungrin/solved/PGM_42861.java
import java.util.*;

public class PGM_42861 {

    public int[] parent;

    public int solution(int n, int[][] costs) {
        int answer = 0;

        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        Arrays.sort(costs, (a, b) ->
                Integer.compare(a[2], b[2])
        );

        int connectCnt = 0;

        for (int[] cost : costs) {
            int a = cost[0];
            int b = cost[1];
            int c = cost[2];

            if (!(find(a) == find(b))) {
                union(a, b);
                answer += c;
                connectCnt++;
            }

            if (connectCnt == n - 1) break;
        }

        return answer;
    }

    private int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    private void union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        parent[rootB] = rootA;
    }
}

File: seoyeong/solved/PGM_42861.py
def solution(n, costs):
    answer = 0
    parent = [0] * n
    
    for i in range(n):
        parent[i] = i
    
    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x        
    
    def union(x, y):
        root_x, root_y = find(x), find(y)
        if root_x == root_y:
            return False
        parent[root_x] = root_y
        return True
    
    costs.sort(key = lambda x: x[2])
    
    used = 0
    for a, b, cost in costs:
        if union(a, b):
            answer += cost
            used += 1
            if used == n-1:
                break
    
    return answer